tftrt/examples/cpp/image-classification/mnist.cc

```diff
@@ -0,0 +1,133 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <algorithm>
+#include <array>
+#include <fstream>
+#include <iostream>
+#include <stdint.h>
+
+#include "mnist.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/lib/core/status.h"
+
+namespace mnist {
+
+namespace {
+
+// This function interprets a 4 byte array as an unsigned,
+// big-endian integer. sizeof(data) must be 4.
+// From Rob Pike's Byte Order Fallacy:
+// https://commandcenter.blogspot.com/2012/04/byte-order-fallacy.html
+uint32_t ConvertBigEndian(unsigned char data[]) {
+  uint32_t result =
+      (data[3] << 0) | (data[2] << 8) | (data[1] << 16) | (data[0] << 24);
+  return result;
+}
+
+} // namespace
+
+void MNISTPrint(const MNISTImage &image) {
+  for (int row = 0; row < MNISTImage::kSize; ++row) {
+    for (int column = 0; column < MNISTImage::kSize; ++column) {
+      std::string pixel = image.buf[row][column] > 0 ? "X" : " ";
+      std::cout << pixel << " ";
+    }
+    std::cout << "\n";
+  }
+}
+
+tensorflow::Tensor MNISTImageToTensor(std::vector<MNISTImage> &images,
+                                      int batch_size) {
+  // https://github.com/tensorflow/tensorflow/issues/8033#issuecomment-520977062
+  tensorflow::Tensor input_image(
+      tensorflow::DT_FLOAT,
+      tensorflow::TensorShape(
+          {batch_size, MNISTImage::kSize, MNISTImage::kSize}));
+  float *img_tensor_flat = input_image.flat<float>().data();
+
+  for (int i = 0; i < batch_size; i++) {
+    std::copy(&images[i].buf[0][0],
+              &images[i].buf[0][0] + (MNISTImage::kSize * MNISTImage::kSize),
+              img_tensor_flat + i * MNISTImage::kSize * MNISTImage::kSize);
+  }
+  return input_image;
+}
+
+MNISTImageReader::MNISTImageReader(const std::string &path)
+    : mnist_path_(path) {}
+
+// MNIST's file format is documented here: http://yann.lecun.com/exdb/mnist/
+// Note(bmzhao): The serialized integers are in big endian, and the magic
+// number is documented to be 2051.
+tensorflow::Status
+MNISTImageReader::ReadMnistImages(std::vector<MNISTImage> *images) {
+  std::ifstream image_file(mnist_path_, std::ios::binary);
+  if (!image_file.is_open()) {
```

Comment:
Nit: I think you can get more types of errors here. We would need to lookup last `errno`, get the string for that, and return it?

Reply:
Yes, I have added that.